lib/Optimizer/Dialect/FIRAttr.cpp

```diff
@@ -0,0 +1,236 @@
+}
+
+UpperBoundAttr UpperBoundAttr::get(mlir::MLIRContext *ctxt) {
+  return AttributeUniquer::get<UpperBoundAttr>(ctxt, getId());
+}
+
+LowerBoundAttr LowerBoundAttr::get(mlir::MLIRContext *ctxt) {
+  return AttributeUniquer::get<LowerBoundAttr>(ctxt, getId());
+}
+
+PointIntervalAttr PointIntervalAttr::get(mlir::MLIRContext *ctxt) {
+  return AttributeUniquer::get<PointIntervalAttr>(ctxt, getId());
+}
+
+// RealAttr
+
+RealAttr RealAttr::get(mlir::MLIRContext *ctxt,
+                       const RealAttr::ValueType &key) {
+  return Base::get(ctxt, getId(), key);
+}
+
+int RealAttr::getFKind() const { return getImpl()->getFKind(); }
+
+llvm::APFloat RealAttr::getValue() const { return getImpl()->getValue(); }
+
+// FIR attribute parsing
+
+namespace {
+mlir::Attribute parseFirRealAttr(FIROpsDialect *dialect,
+                                 mlir::DialectAsmParser &parser,
+                                 mlir::Type type) {
+  int kind;
+  if (parser.parseLess() || parser.parseInteger(kind) || parser.parseComma()) {
+    parser.emitError(parser.getNameLoc(), "expected '<' kind ','");
+    return {};
+  }
+  KindMapping kindMap(dialect->getContext());
+  llvm::APFloat value(0.);
+  if (parser.parseOptionalKeyword("i")) {
+    // `i` not present, so literal float must be present
+    double dontCare;
+    if (parser.parseFloat(dontCare) || parser.parseGreater()) {
+      parser.emitError(parser.getNameLoc(), "expected real constant '>'");
+      return {};
+    }
+    auto fltStr = parser.getFullSymbolSpec()
+                      .drop_until([](char c) { return c == ','; })
+                      .drop_front()
+                      .drop_while([](char c) { return c == ' ' || c == '\t'; })
+                      .take_until([](char c) {
+                        return c == '>' || c == ' ' || c == '\t';
+                      });
+    value = llvm::APFloat(kindMap.getFloatSemantics(kind), fltStr);
+  } else {
+    // `i` is present, so literal bitstring (hex) must be present
+    llvm::StringRef hex;
+    if (parser.parseKeyword(&hex) || parser.parseGreater()) {
+      parser.emitError(parser.getNameLoc(), "expected real constant '>'");
+      return {};
+    }
+    auto bits = llvm::APInt(kind * 8, hex.drop_front(), 16);
+    value = llvm::APFloat(kindMap.getFloatSemantics(kind), bits);
+  }
+  return RealAttr::get(dialect->getContext(), {kind, value});
+}
+} // namespace
+
+mlir::Attribute parseFirAttribute(FIROpsDialect *dialect,
+                                  mlir::DialectAsmParser &parser,
+                                  mlir::Type type) {
+  auto loc = parser.getNameLoc();
+  llvm::StringRef attrName;
+  if (parser.parseKeyword(&attrName)) {
+    parser.emitError(loc, "expected an attribute name");
+    return {};
+  }
+
+  if (attrName == ExactTypeAttr::getAttrName()) {
+    mlir::Type type;
+    if (parser.parseLess() || parser.parseType(type) || parser.parseGreater())
+      parser.emitError(loc, "expected a type");
```

Comment:
Seems like you are missing some returns for these error cases. Lack of coverage?

Reply:
Fixed returns. We need to add negative tests.